Make longjmp over managed frames work

janvorli opened the pull request:
The new exception handling has broken hosting Lua on Windows when a Lua code throws an exception that ends up crossing managed frames. Lua uses longjmp for its exception handling mechanism and the new exception handling reported the longjmp as an unhandled `SEHException`.

This change makes the new EH behave like the legacy one w.r.t. the longjmp propagation over managed frames on Windows. It is a best effort fix though just to get rid of the regression related to the legacy exception handling. New code should not rely on this. And it doesn't work on non-windows targets (neither did it work with the legacy exception handling).

Close #111242

janvorli committed:
Make longjmp over managed frames work

The new exception handling has broken hosting Lua on Windows when
a Lua code throws an exception that ends up crossing managed frames.
Lua uses longjmp for its exception handling mechanism and the new
exception handling reported the longjmp as an unhandled `SEHException`.

This change makes the new EH ignore the `STATUS_LONGJUMP` exception
when it passes through managed frames, so it can reach the target
location. It is a best effort fix though, as finallys in the skipped
frames won't be called.

Close #111242

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -932,6 +932,13 @@ ProcessCLRExceptionNew(IN     PEXCEPTION_RECORD   pExceptionRecord,
 
+    if (pExceptionRecord->ExceptionCode == STATUS_LONGJUMP)
+    {
+        // This is an exception used to unwind during longjmp function. We just let it pass through managed
```
Do we need to unwind Frames to avoid corrupting runtime data structures?

jkotas replied:
And if we are unwinding Frames, is it that hard to call finallys too to match the old behavior?

janvorli replied:
I had it implemented by propagating the exception through the managed frames as usual, so the finallys were called, and then re-issued the longjmp after processing them. The only missing part to do was preventing calling catch handlers. The old EH doesn't call those "naturally", as the longjmp runs only the unwind pass.
And then you've mentioned it was not supported, so I have deleted that code :-).
But anyways, I can change it back to that, the change wasn't complicated. The only thing is that the Unix handling would then be different.

jkotas replied:
> The only thing is that the Unix handling would then be different.

longjump over managed code does not and cannot work well on Unix, so I do not think it is a problem

jkotas commented:
If we want to "support this", should we add a test?

janvorli committed:
Fully simulate legacy EH for longjmp on Windows

janvorli committed:
Fix Unix build

janvorli commented:
/azp run runtime-coreclr outerloop

jkotas commented on src/tests/Regressions/coreclr/GitHub_11242/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+project (Test11242Lib)
```
The issue number is `111242`

janvorli replied:
Thank you for spotting this, fixed.

janvorli committed:
Fix bad merge

janvorli committed:
Fix regression issue id in the test path / name

janvorli commented:
/azp run runtime-coreclr outerloop

jkotas commented:
It may be worth noting that this is best effort for backward compatibility. Skipping over managed frames using setjmp/longjmp is unsupported.

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -8151,6 +8192,13 @@ extern "C" BOOL QCALLTYPE EHEnumNext(EH_CLAUSE_ENUMERATOR* pEHEnum, RhEHClause*
             result = FALSE;
         }
+#ifdef HOST_WINDOWS
+        // When processing longjmp, only finally clauses are considered.
```
This is one of the reasons for not supporting setjmp/longjmp. setjmp/longjmp behaves like uncatchable exception - managed code is not generally robust against uncatchable exceptions.

janvorli replied:
I completely agree. The intent is really just to fix the regression on Windows w.r.t. the legacy EH. It is a best effort and not something people should use in new code. I hope that the fact that it doesn't work on Linux is enough reason for trying to stay away from using longjmp over managed code.

jkotas commented:
LGTM otherwise

janvorli committed:
Add comment explaining this is best effort for back compat

janvorli commented:
> It may be worth noting that this is best effort for backward compatibility. Skipping over managed frames using setjmp/longjmp is unsupported.

I've added a comment to the code and updated the PR description to be explicit about this.

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -5660,14 +5660,30 @@ VOID DECLSPEC_NORETURN DispatchManagedException(OBJECTREF throwable, CONTEXT* pE
+    // On Windows, this enables the possibility to propagate a longjmp across managed frames. Longjmp
+    // behaves like a SEH exception, but only runs the second (unwinding) pass.
+    // NOTE: This is a best effort purely for back compatibility with the legacy exception handling.
+    // It is not guaranteed to work in all cases, and it is not recommended to use it in new code.
```
```suggestion
    // Skipping over managed frames using setjmp/longjmp is
    // is unsupported and it is not guaranteed to work reliably in all cases.
    // https://learn.microsoft.com/dotnet/standard/native-interop/exceptions-interoperability#setjmplongjmp-behaviors
```

jkotas committed:
Update src/coreclr/vm/exceptionhandling.cpp

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -5660,14 +5660,32 @@ VOID DECLSPEC_NORETURN DispatchManagedException(OBJECTREF throwable, CONTEXT* pE
+#ifdef HOST_WINDOWS
+    // On Windows, this enables the possibility to propagate a longjmp across managed frames. Longjmp
+    // behaves like a SEH exception, but only runs the second (unwinding) pass.
+    // NOTE: This is a best effort purely for back compatibility with the legacy exception handling.
```
```suggestion
    // NOTE: This is a best effort purely for backward compatibility with the legacy exception handling.
```

jkotas committed:
Update src/coreclr/vm/exceptionhandling.cpp

jkotas reviewed (approved):
Thank you!

janvorli commented:
@MichalStrehovsky this fix was made just for backward compatibility with the old exception handling, we don't want to fix it for NativeAOT. I am sorry for forgetting disabling the test for NativeAOT.

badrishc commented:
Will this fix make it to the .NET 9 releases? We need it for Lua support in Garnet to handle unknown commands properly.

jkotas commented:
@badrishc It is hard to justify backporting this fix to .NET 9. setjmp/longjmp interop is not supported by .NET: https://learn.microsoft.com/en-us/dotnet/standard/native-interop/exceptions-interoperability#setjmplongjmp-behaviors . This fix does not make it work or supported. It just makes it about as broken as it was, on Windows only. It was and it is still completely broken on Linux. It is not possible to fix that on non-Windows as described in the linked doc. We had a lot of debate whether to bother with fixing this on Windows at all. 

My understanding is that https://github.com/microsoft/Garnet supports both Windows and non-Windows platforms. Is that correct? If it is the case, you will need to avoid depending on setjmp/longjmp interop to make the Lua integration work on non-Windows platforms.

jkotas commented:
The runtime keeps track of the managed/native and native/managed transitions. The transitions are structures allocated on the stack and the runtime Thread data structure keeps pointer (`Thread::m_pFrame`) to the inner-most transition. 

`longjmp` on Unix will unwind the stack, without updating the `m_pFrame` pointer in the Thread structure. It means that it is going to be a dangling pointer for a while that violates the runtime invariants. I guess you are getting lucky in your tests that the stack region where this dangling pointer points to is not overwritten and/or that the dangling pointer gets overwritten with a valid transition pointer before you managed to hit a crash.

kevin-montrose commented:
> `longjmp` on Unix will unwind the stack, without updating the `m_pFrame` pointer in the Thread structure. It means that it is going to be a dangling pointer for a while that violates the runtime invariants. I guess you are getting lucky in your tests that the stack region where this dangling pointer points to is not overwritten and/or that the dangling pointer gets overwritten with a valid transition pointer before you managed to hit a crash.

Well, that's fffffffffff-un 🫠.

I'll spend some more time trying to come up with a design that lets Garnet embed Lua on Linux w/o the possibility of `longjmp`ing over any managed frames.  The tricky bit is Lua uses it for implicit failures, like OOMs - which are hard to avoid, and basically impossible to patch out.

badrishc commented:
> @badrishc It is hard to justify backporting this fix to .NET 9. setjmp/longjmp interop is not supported by .NET: https://learn.microsoft.com/en-us/dotnet/standard/native-interop/exceptions-interoperability#setjmplongjmp-behaviors . This fix does not make it work or supported. It just makes it about as broken as it was, on Windows only. It was and it is still completely broken on Linux. It is not possible to fix that on non-Windows as described in the linked doc. We had a lot of debate whether to bother with fixing this on Windows at all.
> 
> My understanding is that https://github.com/microsoft/Garnet supports both Windows and non-Windows platforms. Is that correct? If it is the case, you will need to avoid depending on setjmp/longjmp interop to make the Lua integration work on non-Windows platforms.

Thanks for the details and context! I want to mention that beyond Garnet, there is fairly wide production use of Lua across Microsoft and outside, and often via the NLua C# library that interops into Lua, and often on Windows. This is also evidenced by @stephentoub's [comment](https://github.com/NLua/NLua/pull/531#issuecomment-2142801625) and [PR](https://github.com/NLua/NLua/pull/531) into NLua last year.

Since Lua uses `longjmp` as its exception handling mechanism (other C libraries might also do the same, perhaps), it would be great to see: (1) how best we can support this on windows, e.g., by backporting this fix to .NET 9; and (2) what, if anything, can .NET do as a _best effort_ to make specifically Lua's style of exception handling work on unix (even though I fully understand your point about it being generally unsupported). Just my 2 cents.
